Added instructions on building hardened images (#14722)
gravitational/teleport: docs/pages/setup/security/hardened-container-images.mdx

---
title: Hardened Container Images
description: How to run Teleport in hardened container images
---

> Customers running in heavily regulated or compliance environments may be

Reviewer: ```suggestion
Customers running in heavily regulated environments with strict compliance requirements may be
```

interested in running Teleport within a minimal or hardened container image.

> This may be customers running in FedRAMP environments or customers that have a

Reviewer 2: ```suggestion
These may be customers running in FedRAMP environments or customers that have a
```

requirement to meet CIS Hardened Image guidelines.

> We have plans to publish harded container images in the second half of 2022.

Reviewer 2: ```suggestion
We have plans to publish hardened container images in the second half of 2022.
```

In the meantime, customers can construct a hardened image that meets those

> requirement by running Teleport within `distroless` container image using the

Reviewer: While distroless may add security due to a reduced surface area, I'm not sure it will actually help meet the requirements you reference here (FedRAMP or CIS). Typically these require you to meet rules for which hardening guidelines are available for, and the orgs that develop these guidelines only target the major OS images.

For example:
- CIS provides a hardened container image that is based on Ubuntu 16.04. 
- DISA has STIGs for [RHEL](https://www.stigviewer.com/stig/red_hat_enterprise_linux_7/) and [Ubuntu](https://ubuntu.com/security/disa-stig)

> example `Dockerfile` below.

Reviewer 2: ```suggestion
example `Dockerfile` below:
```

> ```code

Reviewer 2: ````suggestion
```text
````
Currently the snippet is grayed out. I think `text` would work better here (the docs engine doesn't support Dockerfile syntax highlighting)

# Use multistage builds to extract libgcc_s.so and Teleport binaries.
FROM quay.io/gravitational/teleport:(=teleport.version=) AS teleport-base
FROM gcr.io/distroless/cc-debian11 AS cc

# Run Teleport within a base-debian11 distroless container image.

> FROM gcr.io/distroless/base-debian11

Reviewer 3: Why debian? its one of the distros which typically contains more vulnerabilities. Even the distroless version at the time of writing contains 13 (scanned with trivy).

If the goal is to provide an image more suitable for federal / secure environments, Amazon linux, Alpine, or perhaps even RHEL would be better choices.

Amazon linux and RHEL are also more suited for FIPS.

Could we get install instructions for different distros?

# Copy libgcc_s and Teleport binaries into base-debian11 distroless container.
COPY --from=cc /lib/x86_64-linux-gnu/libgcc_s.so.1 /lib/x86_64-linux-gnu/libgcc_s.so.1
COPY --from=teleport-base /usr/local/bin/teleport /usr/local/bin/teleport
COPY --from=teleport-base /usr/local/bin/tctl /usr/local/bin/tctl
COPY --from=teleport-base /usr/local/bin/tsh /usr/local/bin/tsh
COPY --from=teleport-base /usr/local/bin/tbot /usr/local/bin/tbot

> ```

Reviewer 2: It might be good to add a sentence or two here explaining why a distroless image would be more secure than using the standard images.
